tt-train/sources/ttml/core/distributed_mapping.hpp

```diff
@@ -0,0 +1,279 @@
+
+        // Reshape the list of shards into a 2D grid representing the mesh
+        std::vector<std::vector<xt::xarray<T>>> mesh_shape_tensors;
+        mesh_shape_tensors.reserve(rows);
+        for (int i = 0; i < rows; ++i) {
+            std::vector<xt::xarray<T>> row_tensors;
+            row_tensors.reserve(cols);
+            for (int j = 0; j < cols; ++j) {
+                int index = i * cols + j;
+                row_tensors.push_back(tensors[index]);
+            }
+            mesh_shape_tensors.push_back(std::move(row_tensors));
+        }
+
+        // Concatenate along columns first (within each row)
+        std::vector<xt::xarray<T>> row_concatenated;
+        row_concatenated.reserve(static_cast<size_t>(rows));
+        for (const auto& row : mesh_shape_tensors) {
+            auto concatenated_row = core::concatenate(row, col_dim);
+            row_concatenated.push_back(std::move(concatenated_row));
+        }
+
+        // Then concatenate the resulting tensors along rows
+        auto result = core::concatenate(row_concatenated, row_dim);
+        return result;
+    }
+
+private:
+    tt::tt_metal::distributed::MeshShape m_dims;
+};
+
+template <typename T>
+class ReplicateTensorToMesh : public TensorToMesh<ReplicateTensorToMesh<T>, T> {
+public:
+    using Base = TensorToMesh<ReplicateTensorToMesh<T>, T>;
+    ReplicateTensorToMesh(tt::tt_metal::distributed::MeshShape mesh_shape) : Base(std::move(mesh_shape)) {
+    }
+
+    std::vector<xt::xarray<T>> map_impl(const xt::xarray<T>& tensor) {
+        int num_devices = Base::m_mesh_shape.first * Base::m_mesh_shape.second;
+        std::vector<xt::xarray<T>> tensors;
+        tensors.reserve(static_cast<size_t>(num_devices));
+        for (int i = 0; i < num_devices; ++i) {
+            tensors.push_back(tensor);  // Note: this copies the tensor
+        }
+        return tensors;
+    }
+
+    std::unordered_map<std::string, std::string> config_impl() {
+        int num_devices = Base::m_mesh_shape.first * Base::m_mesh_shape.second;
+        return {{"strategy", "replicate"}, {"replication_factor", std::to_string(num_devices)}};
+    }
+};
+
+template <typename T>
+class ConcatMeshToTensor : public MeshToTensor<ConcatMeshToTensor<T>, T> {
+public:
+    using Base = MeshToTensor<ConcatMeshToTensor<T>, T>;
+    ConcatMeshToTensor(tt::tt_metal::distributed::MeshShape mesh_shape, int dim) :
+        Base(std::move(mesh_shape)), m_concat_dim(dim) {
+    }
+
+    xt::xarray<T> compose_impl(const std::vector<xt::xarray<T>>& tensors) {
+        auto result = core::concatenate(tensors, m_concat_dim);
+        return result;
+    }
+
+private:
+    int m_concat_dim = 0;
+};
+
+template <typename T>
+class VectorMeshToTensor {
+public:
+    VectorMeshToTensor([[maybe_unused]] tt::tt_metal::distributed::MeshShape mesh_shape) {
+    }
+    std::vector<xt::xarray<T>> compose(const std::vector<xt::xarray<T>>& tensors) {
+        return tensors;
+    }
+};
```

Comment:
Let's remove this one? The similar function that live in Python should also be removed.

Reply:
I think its nice to have actually. Like identity mapping.